pubsub/gcloud/aio/pubsub/subscriber.py

```diff
@@ -298,6 +313,15 @@ async def producer(
                 await message_queue.join()
         except asyncio.CancelledError:
             log.info('Producer worker cancelled. Gracefully terminating...')
+
+            if not pull_task.done():
+                # Leaving the connection hanging can result in redelivered
+                # messages, so try to finish before shutting down
+                try:
+                    new_messages += await asyncio.wait_for(pull_task, 5)
```

Comment:
What happens when we wait for a shielded task? Isn't it the same as `await pull_task`?

Reply:
Hmm not sure I understand your question :P Are you talking about the `wait_for` ? I enforced 5 seconds here because we're shutting down and we need time to do that. Or, are you talking about how we're awaiting a shielded task above, and then once cancelled, we await the unshielded task? If you try to await the shielded task again, the `CancelledError` will be raised.

Reply:
Ohh, I see now, we wait for unshielded version of the task, that's why `wait_for` is able to cancel it after 5 seconds! Thank you!

But wait... why does it help the issue? Whenever we close the connection on our side, there is a chance that pubsub will deliver a message to stale connection thus incrementing its `delivery_attempt`. Or do we just increase our odds here so our probability to terminate normally is 5/30?

Reply:
> why does it help the issue

That _is_ the question! I don't have all the details, but I've observed is that we do in fact hit this code path in staging (i.e.  the request is still running after being cancelled). As long as the response is returned within 5 seconds we will successfully recover the otherwise lost messages. Will this work all the time? Probably not. I am guessing it will be more successful under heavy load (i.e. shorter response times). But we shouldn't wait too long if we're going to be terminated soon anyhow. We could make this configurable?

Reply:
Got it. I think making this configurable might be too much, subscriber already has quite a few knobs. I'm happy with how it is now!